### PR TITLE
feat(vmclass): improve vmclass

### PR DESCRIPTION
**Pull request by yaroslavborbat**
## Description
Add features:
* freeze features after first discovery
* add tolerations
* generate available nodes with nodeSelector
* immutable .spec.cpu
* add maxAllocatableResources

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [X] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: crds/doc-ru-virtualmachineclasses.yaml [hunk 1]
```diff
@@ -106,7 +106,23 @@ spec:
                         Карта пар `ключ=значение`. Один `ключ=значение` для matchLabels эквивалентен элементу matchExpressions, ключевым полем которого является «ключ», оператором - «In», а список значений содержит только «значение».
 
                         Для выборки по данному критерию все значения логически суммируются.
-
+                tolerations:
+                  description: |
+                    Параметр аналогичен параметру `spec.tolerations` у [Pod](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/).
+                    Tolerations из этого параметры будут объединены с tolerations, указанными в ресурсе VirtualMachine. При этом tolerations у VirtualMachine имеют больший приоритет.
```
Comment (diafour) [inline crds/doc-ru-virtualmachineclasses.yaml:112]:
```suggestion
                    Tolerations из этого параметра будут объединены с tolerations, указанными в ресурсе VirtualMachine. При этом tolerations у VirtualMachine имеют больший приоритет.
```

File: images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go [hunk 2]
```diff
@@ -67,31 +67,44 @@ func (s *state) VirtualMachines(ctx context.Context) ([]virtv2.VirtualMachine, e
...
+	}
+	var filtered []corev1.Node
+loop:
+	for _, node := range nodes {
+		for _, f := range filters {
+			if f(&node) {
+				continue loop
+			}
+		}
+		filtered = append(filtered, node)
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go:84]: Filter is still inverted.

Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go]: fix

File: images/virtualization-artifact/pkg/controller/common/filter.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+package common
+
+import (
+	"slices"
+)
+
+type FilterFunc[T any] func(obj *T) (keep bool)
+
+func Filter[T any](objs []T, keeps ...FilterFunc[T]) []T {
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/common/filter.go:25]:
```suggestion
// Filter removes item from objs if at least one of the "keeps" functions return false for the item.
func Filter[T any](objs []T, keeps ...FilterFunc[T]) []T {
```

File: images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go [hunk 4]
```diff
@@ -67,31 +67,29 @@ func (s *state) VirtualMachines(ctx context.Context) ([]virtv2.VirtualMachine, e
...
+	var (
+		curr        = s.vmClass.Current()
+		matchLabels map[string]string
+		filters     []common.FilterFunc[corev1.Node]
+	)
 
 	switch curr.Spec.CPU.Type {
 	case virtv2.CPUTypeHost, virtv2.CPUTypeHostPassthrough:
-		return nil, nil
+		// each node
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmclass/internal/state/state.go:87]:
```suggestion
		// Node is always has the "Host" CPU type, no additional filters required.
```

**Review by diafour — APPROVED** [review]

**Pull request merged**